Enables force reasoning based on chat template for Qwen3-Thinking

JustinTong0323 opened the pull request:
<!-- Thank you for your contribution! We appreciate it. The following guidelines will help improve your pull request and facilitate feedback. If anything is unclear, don't hesitate to submit your pull request and ask the maintainers for assistance. -->

## Motivation

Adds a mechanism to enforce reasoning based on patterns detected in the chat template.

The changes introduce a `force_reasoning` flag to the `TemplateManager` and `ReasoningParser`. This flag is set if reasoning patterns are detected in the chat template, such as `<think>` tags.

This ensures that reasoning is enforced even when the user does not explicitly enable it.

<!-- Explain the purpose of this PR and the goals it aims to achieve. -->

## Modifications

<!-- Describe the changes made in this PR. -->

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

JustinTong0323 committed:
Enables force reasoning based on chat template

Adds a mechanism to enforce reasoning based on patterns detected in the chat template.

The changes introduce a `force_reasoning` flag to the `TemplateManager` and `ReasoningParser`. This flag is set if reasoning patterns are detected in the chat template, such as `<think>` tags.

This ensures that reasoning is enforced even when the user does not explicitly enable it.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

zhyncs commented:
After https://github.com/sgl-project/sglang/pull/8363 do we still need this

JustinTong0323 committed:
update

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 commented:
> After #8363 do we still need this在 #8363 之后，我们是否还需要这个

I talked with Qwen team and they think remaining one reasoning parser is better for users, and avoid confusion.

JustinTong0323 committed:
Merge branch 'main' into feat-auto-detect-force-reasoning-qwen3

JustinTong0323 committed:
Unifies Qwen3 reasoning parser

Simplifies the reasoning parser for Qwen3 models by removing the separate "qwen3-thinking" parser.

Now, the "qwen3" parser handles both standard Qwen3 models (with `enable_thinking`) and Qwen3-Thinking models. This simplifies configuration and usage.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
fix none chat-template

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
fix: tokenizer is none

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
Improves reasoning pattern detection

Refines the detection of force reasoning patterns within chat templates
by using regular expressions for more accurate identification.

This change enhances the system's ability to recognize and handle
reasoning prompts, leading to improved model behavior.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
Adds qwen3-thinking reasoning parser option

Adds a dedicated `qwen3-thinking` reasoning parser option.

This commit introduces `qwen3-thinking` as a valid option for the `--reasoning-parser` argument. It allows users to explicitly specify the Qwen3-Thinking model and ensures correct handling of reasoning content, as these models always generate thinking content.

This change clarifies documentation and improves the flexibility and accuracy of reasoning extraction for Qwen3-Thinking models.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
Merge branch 'main' into feat-auto-detect-force-reasoning-qwen3

JustinTong0323 commented:
At present, the behavior is consistent with that of the main branch. Users can choose to use either "qwen3" or "qwen3-thinking". Regarding the Qwen3-Thinking model, these two options exhibit the same behavior.

CatherineSue reviewed (approved):
Overall LGTM. Left two nit comments.

JustinTong0323 committed:
Update docs/backend/openai_api_completions.ipynb

Co-authored-by: Chang Su <csu272@usc.edu>

JustinTong0323 committed:
Update docs/backend/separate_reasoning.ipynb

Co-authored-by: Chang Su <csu272@usc.edu>

JustinTong0323 committed:
Merge branch 'main' into feat-auto-detect-force-reasoning-qwen3

JustinTong0323 committed:
Merge branch 'main' into feat-auto-detect-force-reasoning-qwen3

JustinTong0323 committed:
Merge branch 'main' into feat-auto-detect-force-reasoning-qwen3

JustinTong0323 committed:
feat: add thinking prompt for request in OpenAIServingChat

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
Merge branch 'main' into feat-auto-detect-force-reasoning-qwen3

JustinTong0323 committed:
fix: update thinking token comment and method signature in OpenAIServingChat

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 committed:
Merge branch 'main' into feat-auto-detect-force-reasoning-qwen3

JustinTong0323 committed:
Merge branch 'main' into feat-auto-detect-force-reasoning-qwen3

JustinTong0323 committed:
Merge branch 'main' into feat-auto-detect-force-reasoning-qwen3

JustinTong0323 committed:
fix: disables thinking by default

Updates the default behavior for enabling thinking in chat templates. It now disables reasoning unless explicitly enabled, providing more control over the model's behavior. This change also addresses an issue where reasoning was unintentionally triggered.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
Merge branch 'main' into feat-auto-detect-force-reasoning-qwen3

JustinTong0323 committed:
Merge branch 'main' into feat-auto-detect-force-reasoning-qwen3

JustinTong0323 committed:
Merge branch 'main' into feat-auto-detect-force-reasoning-qwen3

JustinTong0323 committed:
fix: tokenizer is none when skip init

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>
